xds/src/test/java/io/grpc/xds/XdsServerWrapperTest.java

```diff
@@ -0,0 +1,353 @@
+
+@RunWith(JUnit4.class)
+public class XdsServerWrapperTest {
+  @Rule
+  public final MockitoRule mocks = MockitoJUnit.rule();
+
+  @Mock
+  private ServerBuilder<?> mockBuilder;
+  @Mock
+  private Server mockServer;
+  @Mock
+  private static TlsContextManager tlsContextManager;
+  @Mock
+  private XdsServingStatusListener listener;
+
+  private AtomicReference<FilterChainSelector> selectorRef = new AtomicReference<>();
+  private FakeClock executor = new FakeClock();
+  private FakeXdsClient xdsClient = new FakeXdsClient();
+  private XdsServerWrapper xdsServerWrapper;
+
+  @Before
+  public void setup() {
+    when(mockBuilder.build()).thenReturn(mockServer);
+    xdsServerWrapper = new XdsServerWrapper("0.0.0.0:1", mockBuilder, listener,
+            selectorRef, new FakeXdsClientPoolFactory(xdsClient));
+    xdsServerWrapper.setTimeService(executor.getScheduledExecutorService());
+  }
+
+  @Test
+  public void testBootstrap_notV3() throws Exception {
+    Bootstrapper.BootstrapInfo b =
+        new Bootstrapper.BootstrapInfo(
+          Arrays.asList(
+                  new Bootstrapper.ServerInfo("uri", InsecureChannelCredentials.create(), false)),
+          EnvoyProtoData.Node.newBuilder().setId("id").build(),
+          null,
+                "grpc/server?udpa.resource.listening_address=%s");
+    verifyBootstrapFail(b);
+  }
+
+  @Test
+  public void testBootstrap_noTemplate() throws Exception {
+    Bootstrapper.BootstrapInfo b =
+        new Bootstrapper.BootstrapInfo(
+            Arrays.asList(
+                    new Bootstrapper.ServerInfo("uri", InsecureChannelCredentials.create(), true)),
+            EnvoyProtoData.Node.newBuilder().setId("id").build(),
+            null,
+            null);
+    verifyBootstrapFail(b);
+  }
+
+  private void verifyBootstrapFail(Bootstrapper.BootstrapInfo b) throws Exception {
+    XdsClient xdsClient = mock(XdsClient.class);
+    when(xdsClient.getBootstrapInfo()).thenReturn(b);
+    xdsServerWrapper = new XdsServerWrapper("0.0.0.0:1", mockBuilder, listener,
+            selectorRef, new FakeXdsClientPoolFactory(xdsClient));
+    final SettableFuture<Server> start = SettableFuture.create();
+    Executors.newSingleThreadExecutor().execute(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          start.set(xdsServerWrapper.start());
+        } catch (Exception ex) {
+          start.setException(ex);
+        }
+      }
+    });
+    try {
+      start.get(5000, TimeUnit.MILLISECONDS);
+      fail("Start should throw exception");
+    } catch (ExecutionException ex) {
+      assertThat(ex.getCause()).isInstanceOf(IOException.class);
+      Throwable cause = ex.getCause().getCause();
+      assertThat(cause).isInstanceOf(StatusException.class);
+      assertThat(((StatusException)cause).getStatus().getCode())
+              .isEqualTo(Status.UNAVAILABLE.getCode());
+    }
+  }
+
+
+  @Test
+  public void shutdown() throws Exception {
```

Comment:
Do you have tests for the following 2 cases?

- xdsServerWrapper is running but LDS update is "ResourceDoesNotExist" so the delegate server is shutdown. In that state the gRPC app calls `xdsServerWrapper.shutdown()` and that should be handled properly.
- similar to above but after some time LDS update returns a proper listener because of which the code tries to restart delegate server but it fails (because port is in use). So the retryTask attempts to start the delegate server after some interval during which the app calls `xdsServerWrapper.shutdown()`

I am not sure if the code handles it but unit test will catch them